### ISSUE TITLE
Valid User Agent for Mathematica?

### ISSUE THREAD

**Issue by bootylicious1** [BTCMarkets/API#89]
Hi,

It's been brick wall after brick wall for me, trying to get your API to work with Mathematica 9 (because of all the different protections!). I think I am almost there now though because after encoding my public API key to Base-64, instead of receiving the "Authentication failed" response I now receive the following cloudflare response:

```
<head>
<title>Error</title>
</head>
<body>
An error occured. Please contact <a href="/cdn-cgi/l/email-protection" class="__cf_email__" data-cfemail="6d1e181d1d021f192d0f190e000c1f0608191e4303081943">[email&#160;protected]</a>
<script data-cfasync="false" src="/cdn-cgi/scripts/af2821b0/cloudflare-static/email-decode.min.js"></script></body>
</html>
```

I believe this is the response you refer to in you FAQ, which is the result of me not having a valid user agent set in my code.

I am kinda new to using APIs, but I get the impression that this is yer another protection to filter out "adversarial" requests >_<. Could you please tell me what User Agent I should set in my code so I can use you API with Mathematica?

Thank you in advance~

**Comment by meeeee12** [comment]
First off, you don't base64 encode your private key. You need to base64 **decode** you private key before using it, then base64 encode your signature once it has been created. As for your error you are currently seeing, I was also getting that due to a malformed header. Hope this helps.

**Comment by bootylicious1** [comment]
That's what I was doing before >_<

The HMAC SHA-512 Base-64 signatures I generate are consistent with those generated by this website:

https://www.liavaag.org/English/SHA-Generator/HMAC/

The timestamp (which is 13 characters long and in milliseconds) is correct because when i give an invalid one, it says "Authentication failed. invalid timestamp". The timestamp is also consistent with the stamp found on this website:

https://www.epochconverter.com/

I am passing the headers in the correct order ("Accept", "Accept-Charset", "Content-Type", "apikey", "timestamp", "signature"). The public API key I am using for the "apikey" parameter is the same as the one provided on the page https://www.btcmarkets.net/account/apikey. The message used to generate the HmacSHA512Base64 signature is made up of the URI and the time stamp (eg. "/account/balance"<>"\n"<>"1514696341642"), the the time stamp value used in the message is the same value that is used for the "timestamp" parameter, and I am using the GET method to make the request.

What on earth could I be doing wrong?

Please help me sir >_<.

**Comment by meeeee12** [comment]
How are you separating the lines in your header? It should be "\r\n". That was my problem and I was getting the same error, well with GET requests, I would get a different error with POST. Or maybe the other way around, I don't remember anymore.

**Comment by bootylicious1** [comment]
In Mathematica, they are passed as separate things I think. I do it like this.

URLFetch[
 "https://api.btcmarkets.net" <> uri,
 Method -> "GET",
 "Headers" -> {
   "Accept" -> "application/json",
   "Accept-Charset" -> "UTF-8",
   "Content-Type" -> "application/json",
   "apikey" -> publicAPIkey,
   "timestamp" -> timeStamp,
   "signature" -> signature
   }
 ]

I think my version of Mathematica might be old though, so I'm gonna give the new version a go. They have expanded on their html functions, and one of those might do the trick. I'll let you know how it goes >_<.

**Comment by meeeee12** [comment]
I was also doing that (in another language) and it would fail. I switched to another method where I build the header myself, using a string and it worked. Perhaps you also have that option, or maybe set some flag to change the end line character. I have never used Mathematica so I have no idea how.

**Comment by bootylicious1** [comment]
The new line code is "\n" in Mathematica, but I don't think this is the problem since it is receiving the time stamp correctly. I tried adding new lines to each of the parameters anyway just to try, but it didn't work >_<.

I don't know how to build the header myself ;_;

**Comment by meeeee12** [comment]
I was also getting time stamp errors when I deliberately made it incorrect, then I would get the same error as you when I fixed it.

**Comment by bootylicious1** [comment]
Hmm, I think the problem may lie in the formatting of the Hmac message.

Would you be able to tell me the character codes for a valid string requesting the account balance?

For example, in the string:

'/account/balance'+'\n'+'1514696341642'

should I include the ' character in the message, or is that character just something that defines the string boundaries? Should I include \n which has character codes 92 and 100, or is there a separate character in JSON that defines a new line? Also, should I include new line characters at the end?

Any advice would be deeply appreciated.

**Comment by meeeee12** [comment]
you need an extra '\n' on the end of GET requests, so:

'/account/balance'+'\n'+'1514696341642'+'\n'

**Comment by bootylicious1** [comment]
hmm, a step closer, but still no luck ;_;

**Comment by bootylicious1** [comment]
I did it!!

Thank you for helping me kind sir <3.

Turns out, the problem was the missing "\n" at the end. The reason it didn't work earlier is because I turned off Base-64 encoding for the private key while randomly changing things trying to get it to work.

Now all that is left for me to do is get the POST requests working, but that should be a lot easier to figure out now that I know how to do the GET requests!

Thanks again kind sir :3.
